domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,305 @@ class TopicListController @Inject constructor(
       .build()
   }
 
-  private fun createOngoingStoryListFromProgress(
+  private fun createUpcomingTopicSummaryFromJson(
+    topicId: String,
+    jsonObject: JSONObject
+  ): UpcomingTopic {
+    var totalChapterCount = 0
+    val storyData = jsonObject.getJSONArray("canonical_story_dicts")
+    for (i in 0 until storyData.length()) {
+      totalChapterCount += storyData
+        .getJSONObject(i)
+        .getJSONArray("node_titles")
+        .length()
+    }
+    val topicPlayAvailability = if (jsonObject.getBoolean("published")) {
+      TopicPlayAvailability.newBuilder().setAvailableToPlayNow(true).build()
+    } else {
+      TopicPlayAvailability.newBuilder().setAvailableToPlayInFuture(true).build()
+    }
+
+    return UpcomingTopic.newBuilder().setTopicId(topicId)
+      .setName(jsonObject.getString("topic_name"))
+      .setVersion(jsonObject.optInt("version"))
+      .setTopicPlayAvailability(topicPlayAvailability)
+      .setLessonThumbnail(createTopicThumbnail(jsonObject))
+      .build()
+  }
+
+  private fun computePromotedActivityList(
+    topicProgressList: List<TopicProgress>
+  ): PromotedActivityList {
+    val promotedActivityListBuilder = PromotedActivityList.newBuilder()
+    if (topicProgressList.isNotEmpty()) {
+      promotedActivityListBuilder.promotedStoryList = computePromotedStoryList(topicProgressList)
+      if (promotedActivityListBuilder.promotedStoryList.getTotalPromotedStoryCount() == 0) {
+        promotedActivityListBuilder.comingSoonTopicList = computeComingSoonTopicList()
+      }
+    }
+    return promotedActivityListBuilder.build()
+  }
+
+  private fun computePromotedStoryList(topicProgressList: List<TopicProgress>): PromotedStoryList {
+    return PromotedStoryList.newBuilder()
+      .addAllUpTo(
+        populateRecentlyPlayedStories(topicProgressList),
+        PromotedStoryList.Builder::addAllRecentlyPlayedStory,
+        limit = 3
+      )
+      .addAllUpTo(
+        populateOlderPlayedStories(topicProgressList),
+        PromotedStoryList.Builder::addAllOlderPlayedStory,
+        limit = 3
+      )
+      .addAllUpTo(
+        computeSuggestedStories(topicProgressList),
+        PromotedStoryList.Builder::addAllSuggestedStory,
+        limit = 3
+      ).build()
+  }
+
+  private fun PromotedStoryList.getTotalPromotedStoryCount(): Int {
+    return recentlyPlayedStoryList.size + olderPlayedStoryList.size + suggestedStoryList.size
+  }
+
+  private fun PromotedStoryList.Builder.getTotalPromotedStoryCount(): Int {
+    return recentlyPlayedStoryList.size + olderPlayedStoryList.size + suggestedStoryList.size
+  }
+
+  private fun PromotedStoryList.Builder.addAllUpTo(
+    iterable: Iterable<PromotedStory>,
+    addAll: PromotedStoryList.Builder.(Iterable<PromotedStory>) -> PromotedStoryList.Builder,
+    limit: Int
+  ): PromotedStoryList.Builder {
+    return this.addAll(iterable.take(limit - this.getTotalPromotedStoryCount()))
+  }
+
+  private fun populateRecentlyPlayedStories(
```

Comment:
Nit: this isn't populating anything, it's actually computing the list. Suggest updating this & below method names.

Reply:
Removed this method as per the logic https://github.com/oppia/oppia-android/pull/2272#discussion_r560734859